### DISCUSSION TITLE
Would the Rich traceback formatter benefit from having separate exception notes?

### DISCUSSION THREAD

**Discussion by encukou** [Textualize/rich#2049] [Q&A] [answered]
Hello Rich devs,
[PEP 678](https://peps.python.org/pep-0678/) proposes adding *notes* to exceptions in Python 3.11. The abstract says:
> Exception objects are typically initialized with a message that describes the error which has occurred. Because further information may be available when the exception is caught and re-raised, or included in an ExceptionGroup, this PEP proposes to add `BaseException.add_note(note)`, a .`__notes__` attribute holding a tuple of zero or more notes so added, and to update the builtin traceback formatting code to include notes in the formatted traceback following the exception string.
>
> This is particularly useful in relation to [PEP 654](https://peps.python.org/pep-0654) ExceptionGroups, which make previous workarounds ineffective or confusing. Use cases have been identified in the standard library, Hypothesis and cattrs packages, and common code patterns with retries.

As the PEP stands, when several notes are added to an exception, they are collected in a sequence (tuple). There is [some discussion](https://discuss.python.org/t/pep-678-enriching-exceptions-with-notes/13374/56?u=encukou) about this being too complicated: instead of a sequence, there could be a single string where notes would be separated by newlines (but newlines can also appear in individual notes).

How much would Rich's traceback rendering benefit from having the notes separate? How much sadness would there be if they were concatenated?

**Comment by willmcgugan** [comment, marked as answer]
Hi Petr,

My vote would be for the tuple because I can imagine Rich displaying each note within a bullet list. 

I'm sure I could work with it if the notes were concatenated in a str. Ideally there would be some rule or convention that allowed the notes to be separated unambiguously.

Will
